core/src/main/java/org/elasticsearch/rest/action/count/RestCountAction.java

```diff
@@ -59,11 +75,19 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
         CountRequest countRequest = new CountRequest(Strings.splitStringByCommaToArray(request.param("index")));
         countRequest.indicesOptions(IndicesOptions.fromRequest(request, countRequest.indicesOptions()));
         if (RestActions.hasBodyContent(request)) {
-            countRequest.source(RestActions.getRestContent(request));
+            BytesReference restContent = RestActions.getRestContent(request);
+            try (XContentParser requestParser = XContentFactory.xContent(restContent).createParser(restContent)) {
+                QueryParseContext context = new QueryParseContext(indicesQueriesRegistry);
+                context.reset(requestParser);
+                final QueryBuilder<?> builder = context.parseInnerQueryBuilder();
+                countRequest.query(builder);
+            } catch (IOException e) {
+                throw new ElasticsearchException("failed to parse source", e);
+            }
```

Comment:
seems like this whole block could be made a utility method e.g. RestActions.getQueryContent or something along those lines?